lib/zip/file.rb

```diff
@@ -406,7 +405,7 @@ def on_success_replace
       tmp_filename = create_tmpname
       if yield tmp_filename
         ::File.rename(tmp_filename, name)
-        ::File.chmod(@file_permissions, name) if defined?(@file_permissions)
+        ::File.chmod(@file_permissions, name) if @create.nil?
```

Comment:
The main reasons I stopped using the `@file_permissions` variable here are:
* The logic for whether we need to change the permissions here is based upon whether the file is created or not, not whether there exists some file permissions. So,
* If someone decided to use the `@file_permissions` variable for something else, we might get an odd bug here.

I agree with your concerns about the `@create` variable though. I'll look at whether I can make its use consistent.